File: android/app/src/main/java/com/clingsync/android/MergeReminderScheduler.kt
package com.clingsync.android

import android.content.Context
import androidx.work.ExistingWorkPolicy
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.workDataOf
import java.util.Calendar
import java.util.concurrent.TimeUnit

// Runs MergeReminderWorker daily at REMINDER_HOUR local time. A one-time request
// the worker re-arms each day (rather than a periodic one) keeps the trigger
// anchored to the wall-clock hour and re-derived for the current time zone every
// run instead of drifting.
object MergeReminderScheduler {
    const val WORK_NAME = "merge_reminder"
    const val TEST_WORK_NAME = "merge_reminder_test"
    const val REMINDER_HOUR = 18
    const val TEST_DELAY_SECONDS = 8L

    // Arms the reminder only if none is pending. Safe to call on every app start.
    fun ensureScheduled(context: Context) = enqueue(context, ExistingWorkPolicy.KEEP)

    // Re-arms tomorrow's reminder, replacing the run that just finished.
    fun scheduleNext(context: Context) = enqueue(context, ExistingWorkPolicy.REPLACE)

    // Manual test trigger (debug only): runs the worker once in a few seconds,
    // forced onto the daily or weekly path, without touching the real schedule.
    fun scheduleTest(
        context: Context,
        weekly: Boolean,
    ) {
        val mode = if (weekly) MergeReminderWorker.MODE_WEEKLY else MergeReminderWorker.MODE_DAILY
        val request =
            OneTimeWorkRequestBuilder<MergeReminderWorker>()
                .setInitialDelay(TEST_DELAY_SECONDS, TimeUnit.SECONDS)
                .setInputData(workDataOf(MergeReminderWorker.KEY_FORCE_MODE to mode))
                .addTag(TEST_WORK_NAME)
                .build()
        WorkManager.getInstance(context).enqueueUniqueWork(TEST_WORK_NAME, ExistingWorkPolicy.REPLACE, request)
    }

    private fun enqueue(
        context: Context,
        policy: ExistingWorkPolicy,
    ) {
        val request =
            OneTimeWorkRequestBuilder<MergeReminderWorker>()
                .setInitialDelay(millisUntilNextHour(REMINDER_HOUR), TimeUnit.MILLISECONDS)
                .addTag(WORK_NAME)
                .build()
        WorkManager.getInstance(context).enqueueUniqueWork(WORK_NAME, policy, request)
    }

    // Milliseconds from `now` until the next occurrence of `hour`:00 local time.
    fun millisUntilNextHour(
        hour: Int,
        now: Calendar = Calendar.getInstance(),
    ): Long {
        val next = now.clone() as Calendar
        next.set(Calendar.HOUR_OF_DAY, hour)
        next.set(Calendar.MINUTE, 0)
        next.set(Calendar.SECOND, 0)
        next.set(Calendar.MILLISECOND, 0)
        if (!next.after(now)) {
            next.add(Calendar.DAY_OF_MONTH, 1)
        }
        return next.timeInMillis - now.timeInMillis
    }
}

File: ios/ClingSync/MergeReminderScheduler.swift
import BackgroundTasks
import Foundation
import UserNotifications

// Schedules and runs the daily backup reminder via BGTaskScheduler. iOS runs the
// task opportunistically (near, not exactly, the reminder hour); each run
// recomputes whether files are pending and re-arms the next one. The identifier
// must match the one in Info.plist's BGTaskSchedulerPermittedIdentifiers.
enum MergeReminderScheduler {
    static let taskIdentifier = "com.cling.ClingSync.merge-reminder"
    static let reminderHour = 18
    static let testDelaySeconds: TimeInterval = 8

    // Registers the background task handler. Must run before the app finishes
    // launching, so it is called from the app delegate's didFinishLaunching.
    static func registerBackgroundTask() {
        BGTaskScheduler.shared.register(forTaskWithIdentifier: taskIdentifier, using: nil) { task in
            handle(task)
        }
    }

    // Requests notification permission and arms the first reminder. Safe to call on
    // every launch.
    static func start() {
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { _, _ in }
        scheduleNext()
    }

    // Arms the next reminder for the upcoming reminder hour. Resubmitting replaces
    // any pending request for the same identifier.
    static func scheduleNext() {
        let request = BGAppRefreshTaskRequest(identifier: taskIdentifier)
        request.earliestBeginDate = nextReminderDate()
        try? BGTaskScheduler.shared.submit(request)
    }

    // Debug-only trigger (called by the Settings test controls): runs the reminder
    // check in a few seconds, forced onto the daily or weekly path, without touching
    // the real schedule.
    static func scheduleTest(weekly: Bool) {
        Task { await MergeReminderService.runCheck(mode: weekly ? .weekly : .daily, deliverAfter: testDelaySeconds) }
    }

    private static func handle(_ task: BGTask) {
        scheduleNext()
        let work = Task {
            await MergeReminderService.runCheck(mode: .auto)
            task.setTaskCompleted(success: true)
        }
        task.expirationHandler = { work.cancel() }
    }

    private static func nextReminderDate(now: Date = Date(), calendar: Calendar = .current) -> Date {
        var components = calendar.dateComponents([.year, .month, .day], from: now)
        components.hour = reminderHour
        components.minute = 0
        components.second = 0
        let candidate = calendar.date(from: components) ?? now
        if candidate > now {
            return candidate
        }
        return calendar.date(byAdding: .day, value: 1, to: candidate) ?? candidate.addingTimeInterval(24 * 60 * 60)
    }
}
